### PR TITLE
Display submap poses in Submaps plugin

### PR DESCRIPTION
**Pull request by ojura**
This is still unpolished, but I thought I might share this if anyone needs it sooner (or wants to provide feedback). Addresses #405 and #404.

Todo and ideas:
 - toggle axes visibility with a checkbox
 - adjustable length/radius
 - add a selection handler
 - show submap ID next to axes, also toggleable with a checkbox
 - option to display axes even if submap is hidden (for using with the regular occupancy grid map plugin)
 - display numeric pose coordinates under an expandable menu for each submap, similar to TF display

### CODE REVIEW

File: cartographer_rviz/cartographer_rviz/submaps_display.cc [hunk 1]
```diff
@@ -138,8 +138,8 @@ void SubmapsDisplay::processMessage(
       trajectory.emplace(
           submap_index,
           ::cartographer::common::make_unique<DrawableSubmap>(
-              trajectory_id, submap_index, context_->getSceneManager(),
-              trajectory_category.get(), visibility_all_enabled_->getBool()));
+              trajectory_id, submap_index, context_, trajectory_category.get(),
+              visibility_all_enabled_->getBool(), 0.3f, 0.06f));
```
Comment (SirVer) [inline cartographer_rviz/cartographer_rviz/submaps_display.cc]: will we ever want to change these? If so, they should be configurable, if not they should be constexpr next to where they are used. I put my vote in for the second option for now.

File: cartographer_rviz/cartographer_rviz/drawable_submap.h [hunk 2]
```diff
@@ -70,6 +72,10 @@ class DrawableSubmap : public QObject {
   // 'current_tracking_z'.
   void SetAlpha(double current_tracking_z);
 
+  // todo
```
Comment (SirVer) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: they are not implemented. Remove for now and add back when they are actually added?

File: cartographer_rviz/cartographer_rviz/drawable_submap.h [hunk 3]
```diff
@@ -95,11 +101,14 @@ class DrawableSubmap : public QObject {
...
 
   ::cartographer::common::Mutex mutex_;
+  ::rviz::DisplayContext* const context_;
   Ogre::SceneManager* const scene_manager_;
   Ogre::SceneNode* const scene_node_;
+  Ogre::SceneNode* const submap_node_;
   Ogre::ManualObject* manual_object_;
   Ogre::TexturePtr texture_;
   Ogre::MaterialPtr material_;
+  ::rviz::Axes axes_;
```
Comment (SirVer) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: how about `origin_marker_`?

File: cartographer_rviz/cartographer_rviz/drawable_submap.h [hunk 4]
```diff
@@ -46,9 +47,10 @@ class DrawableSubmap : public QObject {
  public:
   // Each submap is identified by a 'trajectory_id' plus a 'submap_index'.
   // 'scene_manager' is the Ogre scene manager to which to add a node.
-  DrawableSubmap(int trajectory_id, int submap_index,
-                 Ogre::SceneManager* scene_manager,
-                 ::rviz::Property* submap_category, const bool visible);
+  DrawableSubmap(const int trajectory_id, const int submap_index,
```
Comment (SirVer) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: const int -> int. The const is an implementation detail for copied and has no meaning in the interface, which (ideally) should be the only thing exposed in the header.

File: cartographer_rviz/cartographer_rviz/drawable_submap.cc [hunk 5]
```diff
@@ -266,15 +273,7 @@ float DrawableSubmap::UpdateAlpha(const float target_alpha) {
...
-  if (visibility_->getBool()) {
-    if (scene_node_->numAttachedObjects() == 0) {
-      scene_node_->attachObject(manual_object_);
-    }
-  } else {
-    if (scene_node_->numAttachedObjects() > 0) {
-      scene_node_->detachObject(manual_object_);
-    }
-  }
+  scene_node_->setVisible(visibility_->getBool());
```
Comment (SirVer) [inline cartographer_rviz/cartographer_rviz/drawable_submap.cc]: Why this change? I vaguely remember you had the longer code before for some performance reasons or so?

File: cartographer_rviz/cartographer_rviz/drawable_submap.h [hunk 6]
```diff
@@ -95,11 +101,14 @@ class DrawableSubmap : public QObject {
   const int submap_index_;
 
   ::cartographer::common::Mutex mutex_;
+  ::rviz::DisplayContext* const context_;
   Ogre::SceneManager* const scene_manager_;
   Ogre::SceneNode* const scene_node_;
+  Ogre::SceneNode* const submap_node_;
```
Comment (SirVer) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: I see that the submap_node_ is made a child node of the scene_node_, but why is the scene_node_ not enough?

File: cartographer_rviz/cartographer_rviz/drawable_submap.cc [hunk 7]
```diff
@@ -55,16 +55,20 @@ std::string GetSubmapIdentifier(const int trajectory_id,
...
+                               ::rviz::Property* const submap_category,
+                               const bool visible, const float axes_length,
+                               const float axes_radius)
     : trajectory_id_(trajectory_id),
       submap_index_(submap_index),
-      scene_manager_(scene_manager),
-      scene_node_(scene_manager->getRootSceneNode()->createChildSceneNode()),
-      manual_object_(scene_manager->createManualObject(
+      context_(context),
+      scene_manager_(context->getSceneManager()),
```
Comment (SirVer) [inline cartographer_rviz/cartographer_rviz/drawable_submap.cc]: delete member? Now you have the context_, you can always get the sceneManager from it?

**Comment by SirVer** [comment]
You are on fire again I see :). Thanks!

Comments are in code.

**Comment by jihoonl** [comment]
Awesome thanks @ojura :smile: :smile: :smile:

File: cartographer_rviz/cartographer_rviz/drawable_submap.h (hunk 3 above)
Comment (ojura) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: It's not a ROS marker. Maybe `pose_axes_`?

File: cartographer_rviz/cartographer_rviz/drawable_submap.cc (hunk 5 above)
Comment (ojura) [inline cartographer_rviz/cartographer_rviz/drawable_submap.cc]: Nah, I just didn't know about `setVisible()`. Plus, now I can handle visibility of both axes and the submap by having them both under a signle `scene_node_`.

File: cartographer_rviz/cartographer_rviz/drawable_submap.h (hunk 6 above)
Comment (ojura) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: See `UpdateTransform()`. `submap_node_` (ex `scene_node_`) has its pose set to the slice pose; I would have to perform additional transformations to obtain the submap origin in the coordinate system of the submap slice. By having a `scene_node_` which has the same pose as the root node, I can just set the axes pose to the one written in `position_` and `orientation_`. Plus, I can hide both the submap and the axes by calling `scene_node_.setVisible()`.

File: cartographer_rviz/cartographer_rviz/submaps_display.cc (hunk 1 above)
Comment (ojura) [inline cartographer_rviz/cartographer_rviz/submaps_display.cc]: Yes, I will add an RViz property for it.

File: cartographer_rviz/cartographer_rviz/drawable_submap.h (hunk 4 above)
Comment (ojura) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: Okay. And the boolean variable retains the const in the header?

Comment (SirVer) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: No, that was wrong already. Could you fix it?

File: cartographer_rviz/cartographer_rviz/submaps_display.cc (hunk 1 above)
Comment (SirVer) [inline cartographer_rviz/cartographer_rviz/submaps_display.cc]: next PR?

File: cartographer_rviz/cartographer_rviz/drawable_submap.h (hunk 2 above)
Comment (SirVer) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: There is no such thing as a too small PR. Next PRs?

File: cartographer_rviz/cartographer_rviz/drawable_submap.h (hunk 3 above)
Comment (SirVer) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: fair enough. pose_axes_ is a good name.

File: cartographer_rviz/cartographer_rviz/drawable_submap.h (hunk 4 above)
Comment (ojura) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: Okay, sure. The header file is full of consts though. Should I remove them all?

Comment (SirVer) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: I did that real quick in this branch while reviewing.

**Comment by SirVer** [comment]
@ojura I rebased while having another quick look over this.

File: cartographer_rviz/cartographer_rviz/drawable_submap.h (hunk 4 above)
Comment (ojura) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: Sure. The rviz properties pointers should also have the `const` removed, right?

Comment (SirVer) [inline cartographer_rviz/cartographer_rviz/drawable_submap.h]: Yes... I only did half of the job :(

**Comment by ojura** [comment]
@SirVer okay, I addressed the nits, it's not yet finished (I added a TODO for making it configurable from RViz) but it should be ready for merging :)

**Comment by ojura** [comment]
Rebased to follow #419.

**Comment by ojura** [comment]
Sure.

**Comment by SirVer** [comment]
Thanks for your work and sorry that this took so long on my side.
